src/ngraph/runtime/gpu/gpu_external_function.cpp

```diff
@@ -440,102 +373,67 @@ using namespace std;
                 writer << emit_string_array(c_value_strings, 100 - writer.indent * 4);
                 writer.indent--;
                 writer << "\n};\n\n";
-                writer << "static " << tv->get_tensor().get_element_type().c_type_string() << " *"
-                       << tv->get_tensor().get_name() << ";\n";
                 m_variable_name_map[tv->get_tensor().get_name()] = tv->get_tensor().get_name();
             }
         }
     }
     // Add cudnn descriptor factory for descriptor management.
     // After the cuDNN code emitted in gpu_emitter.cc is refactored
     // into the CUDNNEmitter class, this can be removed.
-    writer << "static runtime::gpu::CUDNNDescriptors descriptors;\n";
+    writer << "static runtime::gpu::CUDNNDescriptors descriptors;\n\n";
 
     writer << "// Declare all functions\n";
     for (shared_ptr<Function> f : pass_manager.get_state().get_functions())
     {
         writer << "extern \"C\" void " << f->get_name() << "(void** inputs, void** outputs, "
                << "gpu::GPURuntimeContext* ctx);\n";
     }
-
     writer << "\n";
 
-    unordered_map<Node*, string> match_functions;
+    // This for loop creates a collection of functions that are called more than once
+    // and emitting them as globally callable functions.
+    // ops implement the is_functionally_identical method
+    unordered_map<string, string> match_function_map;
+    unordered_map<const Node*, string> node_function_map;
     for (shared_ptr<Function> current_function : pass_manager.get_state().get_functions())
     {
-        set<string> output_names;
-        for (shared_ptr<Node> op : current_function->get_results())
-        {
-            shared_ptr<descriptor::TensorView> tv = op->get_output_tensor_view();
-            output_names.insert(tv->get_tensor().get_name());
-        }
-        const list<shared_ptr<Node>>& tmp = current_function->get_ordered_ops();
+        list<shared_ptr<Node>> tmp = current_function->get_ordered_ops();
         if (tmp.size() < 2)
         {
             // Since we are comparing ops there must be at least two ops to proceed.
             continue;
         }
         vector<shared_ptr<Node>> op_list{tmp.begin(), tmp.end()};
-        for (size_t i = 0; i < op_list.size() - 1; i++)
+        for (size_t i = 0; i < op_list.size(); i++)
         {
             if (op_list[i]->is_constant() || op_list[i]->is_parameter())
             {
                 continue;
             }
-            if (contains_key(match_functions, op_list[i].get()))
+
+            Node& node = *op_list[i];
+            auto handler = dispatcher.find(type_index(typeid(node)));
+            if (handler == dispatcher.end())
             {
-                continue;
+                throw ngraph_error("Unhandled op during code generation : " + node.description());
             }
+
+            string match_function = emit_op_as_function(node, "__f__");
```

Comment:
It can't be const because the arg memory manager is done in the emit function. I agree with you that it SHOULD be const, but until the arg memory manager is done is a pass this has to be non-const.

Reply:
Got it, 👍